libp2p/protocols/secure/noise.nim

```diff
@@ -146,15 +146,16 @@ proc encrypt(
 proc encryptWithAd(
     state: var CipherState, ad, data: openArray[byte]
 ): seq[byte] {.raises: [NoiseNonceMaxError].} =
-  result = newSeqOfCap[byte](data.len + sizeof(ChaChaPolyTag))
-  result.add(data)
+  let tag = encrypt(state, data, ad)
 
-  let tag = encrypt(state, result, ad)
-
-  result.add(tag)
+  var res = newSeqUninit[byte](data.len + tag.len)
+  copyMem(addr res[0], unsafeAddr data[0], data.len)
```

Comment:
something to be careful with is that when `data.len == 0`, `[0]` will raise a defect. not sure if we can end with that here in particular, but it's a simple mistake to make.